pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -174,24 +179,23 @@ func getTierPriority(tier string) networking.TierPriority {
 	return tierPriorityMap[tier]
 }
 
-func (n *NetworkPolicyController) toAntreaPeerForCRD(peers []secv1alpha1.NetworkPolicyPeer, cnp *secv1alpha1.ClusterNetworkPolicy, dir networking.Direction) *networking.NetworkPolicyPeer {
+func (n *NetworkPolicyController) toAntreaPeerForCRD(peers []secv1alpha1.NetworkPolicyPeer, cnp *secv1alpha1.ClusterNetworkPolicy, dir networking.Direction, namedPortExists bool) *networking.NetworkPolicyPeer {
 	var addressGroups []string
 	// Empty NetworkPolicyPeer is supposed to match all addresses.
 	// It's treated as an IPBlock "0.0.0.0/0".
 	if len(peers) == 0 {
-		// For an ingress Peer, skip adding the AddressGroup matching all Pods
-		// because in case of ingress Rule, the named Port resolution happens on
-		// Pods in AppliedToGroup.
-		if dir == networking.DirectionIn {
+		// For an egress Peer that specifies any named ports, it creates or
+		// reuses the AddressGroup matching all Pods in all Namespaces and
+		// appends the AddressGroup UID to the returned Peer such that it can be
+		// used to resolve the named ports.
+		// For other cases it uses the IPBlock "0.0.0.0/0" to avoid the overhead
+		// of handling member updates of the AddressGroup.
+		if dir == networking.DirectionIn || !namedPortExists {
 			return &matchAllPeer
 		}
-		// For an egress Peer, create an AddressGroup matching all Pods in all
-		// Namespaces such that it can be used to resolve named Ports. This
-		// AddressGroup is set in the NetworkPolicyPeer of matchAllPeer.
 		allPodsGroupUID := n.createAddressGroupForCRD(matchAllPodsPeerCrd, cnp)
 		podsPeer := matchAllPeer
-		addressGroups = append(addressGroups, allPodsGroupUID)
-		podsPeer.AddressGroups = addressGroups
+		podsPeer.AddressGroups = append(addressGroups, allPodsGroupUID)
```

Comment:
Do you think we can further optimize to include only Pods with the specified namedPorts, or at least do not include Pods without any namedPorts?

Reply:
+1 to this comment. Although maybe there is a trade off here, as we may end up with multiple large address groups instead of one large one?

Reply:
If we want to `include only Pods with the specified namedPorts`, one proposal is to add capabilities in addressGroups to select pods with certain namedPort. Right now an addressGroup selector is only defined by namespaceSelector and podSelector (externalEntitySelector soon).  With the new 'namedPort selector', we can do the namedPort resolution in filterAddressGroupsForPod(). However I believe this fundamentally changes how namedPort resolution is done right now and would require quite some efforts? @tnqn 

Another thing that confuses me is that, if there is named port in egress rule, why do we append 0.0.0.0/24? This is essentially saying, for out of cluster traffic, we simply ignore the namedPort rather than interpreting it as resolving to nothing? Is this is expected semantics?

Reply:
Looks like github sorts comments based on "drafted time" instead of published time. (Mine is actually later than @Dyanngg's comment)

@Dyanngg https://github.com/vmware-tanzu/antrea/pull/1100#discussion_r472670355 should answer your first question, at the moment I don't feel it's worth to do this change given the listed reasons.

For the second question, I think you mean "0.0.0.0/0".
```
	// List of destinations for outgoing traffic of pods selected for this rule.
	// Items in this list are combined using a logical OR operation. If this field is
	// empty or missing, this rule matches all destinations (traffic not restricted by
	// destination). If this field is present and contains at least one item, this rule
	// allows traffic only if the traffic matches at least one item in the to list.
	// +optional
	To []NetworkPolicyPeer `json:"to,omitempty" protobuf:"bytes,2,rep,name=to"`
```
When "To" is not specified, it means allowing all destinations, so "0.0.0.0/0" is a must, we prepend the "matchAllPods" Group is to support named port case, see this test case https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/network_policy.go#L691.
If we don't do it, the named port would be ignored as "0.0.0.0/0" cannot resolve any port name, but apparently it's expected that the name should be converted number if a Pod is able to.

Reply:
We discussed to include NamedPorts into group definition before. I know it will add much complexity and probably need more thinking.

For a global group to include all Pods without any NamedPort, I still feel it is doable, though yes we have to handle all different cases and judge which cases can use this group. I think it also depends on what kinds of rules are more common (e.g. if rules likely contain both NamedPorts and un-named ports). If we are not clear of these for now, probably let us see if we can get more feedback later.

Reply:
in general i feel the named port in network policy is not a highly used use case . so i feel adding the additional complexity to handle the named ports is probably not worth doing. but as Jianjun says, we need to see how people using antrea use the network policies ... so we can decide a better approach when we have more data

Reply:
@jianjuns @abhiraut sure, let's consider this when there are cases proving it's worth.